### DISCUSSION TITLE
OS.js BrowserFS Adapter

### DISCUSSION THREAD

**Discussion by saoirse-iontach** [os-js/OS.js#857] [Show and tell]
<img align="right" src="https://www.tutorialkart.com/wp-content/uploads/2017/09/node-fs.png" height=70>
<img align="right" src="https://raw.githubusercontent.com/os-js/gfx/master/logo-footer.png" height=45>

# OS.js BrowserFS Adapter

This is a BrowserFS VFS Adapter for OS.js.

[OS.js](https://www.os-js.org/) is an [open-source](https://raw.githubusercontent.com/os-js/OS.js/master/LICENSE) web desktop platform with a window manager, application APIs, GUI toolkit, filesystem abstractions and much more.

[@BrowserFS](https://github.com/saoirse-iontach/browser-fs-core) is an in-browser file system that emulates the [Node JS file system API](http://nodejs.org/api/fs.html) and supports storing and retrieving files from various backends. BrowserFS also integrates nicely with other tools.

[![Home](https://img.shields.io/badge/github-repository-1f883d.svg)](https://github.com/saoirse-iontach/osjs-browserfs-adapter) &nbsp;&nbsp; [![Test](https://img.shields.io/badge/github-test_page-92d992.svg)](https://saoirse-iontach.github.io/osjs-browserfs-adapter/_index.html) &nbsp;&nbsp; <sup>[(- saoirse-iontach -)](https://github.com/saoirse-iontach)</sup>

**Comment by coder0107git** [comment]
I'm a little confused. The `@BrowserFS` link takes you to your fork of ZenFS. What are the advantages of using your fork versus ZenFS or the old deprecated BrowserFS?

**Reply by saoirse-iontach** [reply]
Yes, it was very confusing for me too. James P claimed his @BrowserFS was the new browserfs ; At this time, it was mainly a fork, but still many unreliable changes, and that's why I have to fork and fix it. Some month later, he rebrand it as ZenFS and diverge a lot, an still not working.... (Still exist [@BrowserFS](https://github.com/browser-fs) organisation, but repositories have been moved to ZenFS)

Note: **Don't confuse** James P. BrowserFS/ZenFS with other existing homonyms.
[BrowserFS](https://github.com/jvilk/BrowserFS) is a project of **John Vilk**.
[ZenFS](https://github.com/westerndigitalcorporation/zenfs) is a **Western Digital** file system plugin for [RocksDB](https://zonedstorage.io/docs/applications/zenfs).
- On npm:
  - [BrowserFS](https://www.npmjs.com/package/browserfs) is a [jvilk](https://www.npmjs.com/~jvilk) package in npm
  - [@browser-fs](https://www.npmjs.com/org/browser-fs) is an empty npm organization
  - @BrowserFS **was** a npm organization of James P.
  - [@ZenFS](https://www.npmjs.com/org/zenfs) **is** the new npm organization of James P.
- On github:
  - [browserfs](https://github.com/browserfs) and [zenfs](https://github.com/zenfs) are no interest organisations or users
  - [browser-fs](https://github.com/browser-fs) and [zen-fs](https://github.com/zen-fs) are James P. organisations

_Yes, inconsistent names. That's what happens when you want to take advantage of the reputation of existing projects._

ZenFS should be avoid for now, and I will have to rollback to jVilk/browserfs.
In the meantime, my @BrowserFS fork is stable an working (storage and FSAccess).
(I have not tested nor fixed other backends)

Moreover @BrowserFS breaks compatibility with jVilk/browserfs, so that for now we can't reuse existing backends. Also it was a little insane to manage AMD and WebPack support because of split between Core and DOM ; I could have concatenated them, but initial intent was to work with requirejs and UNPKG to build a modular standalone OS-js.

As this 3 projects expose the same node:fs API, with some little implementation details, it should be easy to patch this adapter to work we any of them.

Main difference between old and new is :
 - in jVilk/browserfs we interface with instantiable MountableFileSystem backend.
 - in jamesP @BrowserFS we interface with singleton emulation modules.

Also I will have to investigate to found a reliable base version from github:jVilk/browserfs, because James P have done many change there. I think the reference should be the version published to npm, because people usually install from npm registry, but there are some fix in github.

**Reply by james-pre** [reply]
Hey @coder0107git, I just found this discussion and would like to address some inconsistent and incorrect information that was given to you.

> ... [redacted] claimed his @&#8203;BrowserFS was the new browserfs ; At this time, it was mainly a fork, but still many unreliable changes, ...

At the time, I was (and still am, to a degree), the current and official maintainer of BrowserFS. The decision to turn the majority of my work into a fork was based on the difficulty of getting changes out to the users. As jvilk was extremely busy, he could not administer to BrowserFS. Consequently, it was not possible for me to publish releases for the `browserfs` NPM package. I felt my changes could not be widely used, and opted to instead publish my work as a fork.

> Note: Don't confuse [redacted] BrowserFS/ZenFS with other existing homonyms.

For clarification, @browser-fs and the NPM organization `browserfs` were my efforts to publish my changes. Unfortunately, this resulted in a lot of confusion from the community over whether it was official or not. Despite my continued assurances as to the legitimacy of the work, it appears it fell on deaf ears in many cases. This provided additional motivation for me to publish my work as a fork.

> That's what happens when you want to take advantage of the reputation of existing projects.

This is libel. I never attempted to "take advantage" of the reputation of BrowserFS. In fact, I spent hundreds of hours putting work in to modernize the library. Only when I noticed confusion did I rebrand my work.

> ZenFS should be avoid[ed] for now ...

There is no factual basis for this statement.

> In the meantime, my @BrowserFS fork is stable an[d] working (storage and FSAccess).

My work is absolutely stable, even releasing a 1.0 release at the start of this month. Furthermore, my work includes significant bug fixes not present in the fork of my fork.

I hope I was able to clarify the situation. Please let me know if you'd like any more information.
